### PR TITLE
MachineVerifier: Reject extra non-register operands on instructions

### PR DESCRIPTION
**Pull request by arsenm**
We were allowing extra immediate arguments, and only bothering to check if registers were implicit or not.

Also consolidate extra operand checks in verifier, to make this testable. We had 3 different places checking if you were trying to build an instruction with more operands than allowed by the definition. We had an assertion in addOperand, a direct check in the MIRParser to avoid the assertion, and the machine verifier checks. Remove the assert and parser check so the verifier can provide a consistent verification experience, which will also handle instructions modified in place.

### CODE REVIEW

File: llvm/lib/CodeGen/MIRParser/MIParser.cpp [hunk 1]
```diff
@@ -1175,19 +1175,10 @@ bool MIParser::parse(MachineInstr *&MI) {
...
-    bool IsImplicitOp = Operand.Operand.isReg() && Operand.Operand.isImplicit();
-    if (!IsImplicitOp) {
-      if (!MCID.isVariadic() && NumExplicitOps >= MCID.getNumOperands() &&
-          !Operand.Operand.isValidExcessOperand())
-        return error(Operand.Begin, "too many operands for instruction");
-
-      ++NumExplicitOps;
-    }
-
+  // Don't check the operands make sense, let the verifier catch any
```
Comment (kparzysz) [inline llvm/lib/CodeGen/MIRParser/MIParser.cpp:1178]: I think there is "if" missing in the comment.

File: llvm/lib/CodeGen/MachineInstr.cpp [hunk 2]
```diff
@@ -231,10 +231,6 @@ void MachineInstr::addOperand(MachineFunction &MF, const MachineOperand &Op) {
   // OpNo now points as the desired insertion point.  Unless this is a variadic
   // instruction, only implicit regs are allowed beyond MCID->getNumOperands().
   // RegMask operands go between the explicit and implicit operands.
-  assert((MCID->isVariadic() || OpNo < MCID->getNumOperands() ||
-          Op.isValidExcessOperand()) &&
-         "Trying to add an operand to a machine instr that is already done!");
-
```
Comment (qcolombet) [inline llvm/lib/CodeGen/MachineInstr.cpp:237]: I feel that assert still has value to actually catch when we screw up the operation creation.

File: llvm/test/CodeGen/MIR/AMDGPU/extra-reg-operand.mir [hunk 3]
```diff
@@ -1,13 +0,0 @@
...
-# RUN:   FileCheck --strict-whitespace %s
-
----
-name: extra_reg_operand
-body: |
-  bb.0:
-    ; CHECK: [[@LINE+3]]:17: too many operands for instruction
-    ; CHECK-NEXT: {{^}}    S_ENDPGM 0, undef $vgpr0
-    ; CHECK-NEXT: {{^}}                ^
-    S_ENDPGM 0, undef $vgpr0
```
Comment (qcolombet) [inline llvm/test/CodeGen/MIR/AMDGPU/extra-reg-operand.mir:11]: The message may have changed, but the test may still be valuable, isn't it?

**Review by kparzysz — COMMENTED** [review]
Does the verifier run even when it's not explicitly enabled?

Otherwise this makes sense to me.

File: llvm/lib/Target/MSP430/MSP430RegisterInfo.cpp [hunk 4]
```diff
@@ -134,6 +134,7 @@ MSP430RegisterInfo::eliminateFrameIndex(MachineBasicBlock::iterator II,
 
     MI.setDesc(TII.get(MSP430::MOV16rr));
     MI.getOperand(FIOperandNum).ChangeToRegister(BasePtr, false);
+    MI.removeOperand(FIOperandNum + 1);
```
Comment (qcolombet) [inline llvm/lib/Target/MSP430/MSP430RegisterInfo.cpp:137]: May be add a comment: 'Get rid of the unused Offset operand'

**Review by qcolombet — APPROVED** [review]
LGTM, nits and suggestions below

**Review by asl — APPROVED** [review]
LGTM, thanks!

File: llvm/test/CodeGen/MIR/AMDGPU/extra-reg-operand.mir (hunk 3 above)
Comment (arsenm) [reply llvm/test/CodeGen/MIR/AMDGPU/extra-reg-operand.mir:11]: This should now be covered by the verifier tests

File: llvm/lib/CodeGen/MIRParser/MIParser.cpp (hunk 1 above)
Comment (jayfoad) [reply llvm/lib/CodeGen/MIRParser/MIParser.cpp:1178]: It reads OK to me. "Don't check the operands make sense" is a shorter way of saying "Don't check _that_ the operands make sense".

**Comment by arsenm** [comment]
> Does the verifier run even when it's not explicitly enabled?

For MIR parsing it does. The enable option only changes whether it runs between passes

File: llvm/lib/CodeGen/MIRParser/MIParser.cpp (hunk 1 above)
Comment (kparzysz) [reply llvm/lib/CodeGen/MIRParser/MIParser.cpp:1178]: I somewhat misparsed it the first time I read it, and then it looked weird to me every time I looked at it.  :)

**Pull request merged**